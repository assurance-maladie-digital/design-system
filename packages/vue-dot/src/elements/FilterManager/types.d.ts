import { FilterItem } from '../../patterns/FilterModule/types';

export interface FilterManagerItem {
	index: number,
	filter: FilterItem
}
