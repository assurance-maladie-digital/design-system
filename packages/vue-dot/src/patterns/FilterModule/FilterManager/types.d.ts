import { Field } from '@cnamts/form-builder/src/components/FormField/types';

export interface FilterManagerItem {
	index: number;
	filter: Field;
}
