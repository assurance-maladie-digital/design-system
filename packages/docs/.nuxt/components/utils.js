// nuxt/nuxt.js#8607
export function wrapFunctional(options) {
  if (!options || !options.functional) {
    return options
  }

  const propKeys = Array.isArray(options.props) ? options.props : Object.keys(options.props || {})

  return {
    render(h) {
      const attrs = {}
      const props = {}

      for (const key in this.$attrs) {
        if (propKeys.includes(key)) {
          props[key] = this.$attrs[key]
        } else {
          attrs[key] = this.$attrs[key]
        }
      }

      return h(options, {
        on: this.$listeners,
        attrs,
        props,
        scopedSlots: this.$scopedSlots,
      }, this.$slots.default)
    }
  }
}
