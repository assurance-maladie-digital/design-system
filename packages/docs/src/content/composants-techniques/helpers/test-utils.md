---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
