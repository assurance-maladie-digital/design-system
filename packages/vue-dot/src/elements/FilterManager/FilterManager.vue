<template>
	<VMenu offset-y>
		<template #activator="{ on, attrs }">
			<VBtn
				color="secondary"
				dark
				rounded
				v-bind="attrs"
				class="mr-2"
				style="text-transform:lowercase;"
				v-on="on"
			>
				{{ displayFiltersCount }}
				<VIcon
					color="white"
					class="ml-2"
					@click="resetFilters"
				>
					{{ closeIcon }}
				</VIcon>
			</VBtn>
		</template>
		<VList>
			<VListItem
				v-for="(item, index) in displayFiltersRow"
				:key="index"
			>
				<VListItemTitle>
					<div style="display: flex; align-items:center;">
						<div style="width:300px;">
							<div class="my-2">
								{{ item.filter.label }}
							</div>
							<div class="my-2">
								{{ `${item.filter.form.filter.value.value.length} ${item.filter.label} ${item.filter.form.filter.value.value.length > 1 ? 'sélectionnés' : 'sélectionné'}` }}
							</div>
						</div>
						<div
							class="mr-2"
							@click="editFilters(item.index)"
						>
							<VIcon
								color="black"
								class=""
							>
								{{ editIcon }}
							</VIcon>
						</div>
						<div
							class="mr-2"
							@click="clearFiltersRow(item.index)"
						>
							<VIcon
								color="black"
								class=""
							>
								{{ deleteIcon }}
							</VIcon>
						</div>
					</div>
				</VListItemTitle>
			</VListItem>
		</VList>
	</VMenu>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { mdiCloseCircle } from '@mdi/js';
	import { mdiWindowClose } from '@mdi/js';
	import { mdiPencil } from '@mdi/js';

	import { FilterItem } from '../../patterns/FilterModule/types';
	import { FilterManagerItem } from './types';

	const Props = Vue.extend({
		props: {
			appliedFilters: {
				type: Array as PropType<FilterItem[]>,
				required: true
			}
		}
	});

	const MixinsDeclaration = mixins(Props);

	@Component
	export default class FilterManager extends MixinsDeclaration {
		// Icon
		closeIcon = mdiCloseCircle;
		deleteIcon = mdiWindowClose;
		editIcon = mdiPencil;

		editFilters(index: number): void {
			this.$emit('edit-filters', index);
		}

		clearFiltersRow(index: number): void {
			this.$emit('clear-filters-row', index);
		}

		resetFilters(): void {
			this.$emit('reset-filters');
		}

		get displayFiltersRow(): FilterManagerItem[] {
			const displayedFilters: FilterManagerItem[] = [];
			this.appliedFilters.forEach((filter, index) => {
				let item = {
					index: index,
					filter: filter
				};
				filter.form.filter.value ? displayedFilters.push(item) : '';
			});
			return displayedFilters;
		}

		get displayFiltersCount(): string {
			return `${this.displayFiltersRow.length} ${this.displayFiltersRow.length > 1 ? 'filtres' : 'filtre'}`;
		}
	}
</script>

<style lang="scss" scoped>

</style>
