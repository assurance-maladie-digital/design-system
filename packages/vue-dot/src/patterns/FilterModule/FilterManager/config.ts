export const config = {
	menu: {
		offsetY: true
	},
	editBtn: {
		icon: true
	},
	editIcon: {
		color: 'grey darken-1'
	},
	deleteBtn: {
		icon: true
	},
	deleteIcon: {
		color: 'grey darken-1'
	},
	closeBtn: {
		color: 'secondary',
		rounded: true,
		class: 'text-lowercase mr-2'
	},
	closeIcon: {
		color: 'white',
		class: 'ml-2'
	}
};
