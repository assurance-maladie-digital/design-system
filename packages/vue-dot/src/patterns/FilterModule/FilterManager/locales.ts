export const locales = {
	filter: (plural: boolean): string => `filtre${plural ? 's' : ''}`
};
