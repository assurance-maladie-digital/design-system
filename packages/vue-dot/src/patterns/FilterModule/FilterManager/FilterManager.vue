<template>
	<VMenu
		v-bind="options.menu"
		attach
	>
		<template #activator="{ on, attrs }">
			<VBtn
				v-bind="{
					...attrs,
					...options.closeBtn
				}"
				v-on="on"
			>
				{{ filtersCount }}

				<VIcon
					v-bind="options.closeIcon"
					@click="resetFilters"
				>
					{{ closeIcon }}
				</VIcon>
			</VBtn>
		</template>

		<VList v-bind="options.list">
			<VListItem
				v-for="(item, index) in filters"
				:key="index"
				v-bind="options.listItem"
			>
				<div class="d-flex align-center">
					<div class="vd-filter-manager my-2">
						{{ item.filter.fieldOptions.filterTitle }}
					</div>

					<VBtn
						v-bind="options.editBtn"
						@click="editFilter(item.index)"
					>
						<VIcon v-bind="options.editIcon">
							{{ editIcon }}
						</VIcon>
					</VBtn>

					<VBtn
						v-bind="options.deleteBtn"
						@click="clearFilter(item.index)"
					>
						<VIcon v-bind="options.deleteIcon">
							{{ deleteIcon }}
						</VIcon>
					</VBtn>
				</div>
			</VListItem>
		</VList>
	</VMenu>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { config } from './config';
	import { locales } from './locales';

	import { customizable } from '../../../mixins/customizable';

	import { mdiCloseCircle, mdiWindowClose, mdiPencil } from '@mdi/js';

	import { Field } from '@cnamts/form-builder/src/components/FormField/types';
	import { FilterManagerItem } from './types';

	const Props = Vue.extend({
		props: {
			appliedFilters: {
				type: Array as PropType<Field[]>,
				required: true
			}
		}
	});

	const MixinsDeclaration = mixins(Props, customizable(config));

	@Component
	export default class FilterManager extends MixinsDeclaration {
		locales = locales;

		closeIcon = mdiCloseCircle;
		deleteIcon = mdiWindowClose;
		editIcon = mdiPencil;

		get filters(): FilterManagerItem[] {
			const displayedFilters: FilterManagerItem[] = [];

			this.appliedFilters.forEach((filter, index) => {
				const item = {
					index,
					filter
				};

				if (filter.value) {
					displayedFilters.push(item);
				}
			});

			return displayedFilters;
		}

		get filtersCount(): string {
			const plural = this.filters.length > 1;

			return `${this.filters.length} ${locales.filter(plural)}`;
		}

		editFilter(index: number): void {
			this.$emit('edit-filter', index);
		}

		clearFilter(index: number): void {
			this.$emit('clear-filter', index);
		}

		resetFilters(): void {
			this.$emit('reset-filters');
		}
	}
</script>

<style lang="scss" scoped>
	.vd-filter-manager {
		width: 300px;
	}
</style>
