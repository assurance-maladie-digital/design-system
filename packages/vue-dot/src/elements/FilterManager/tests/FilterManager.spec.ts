import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';
import { html } from '@/tests/utils/html';

import FilterManager from '../';

let wrapper: Wrapper<Vue>;

// Tests
describe('FilterManager', () => {
	it('renders correctly', () => {
		// Mount component
		wrapper = mountComponent(FilterManager);

		expect(html(wrapper)).toMatchSnapshot();
	});
});
