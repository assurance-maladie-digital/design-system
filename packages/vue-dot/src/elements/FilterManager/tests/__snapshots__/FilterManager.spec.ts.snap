// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterManager renders correctly 1`] = `
<vmenu-stub opendelay="0" closedelay="0" contentclass="" maxwidth="auto" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" openonclick="true" closeonclick="true" closeoncontentclick="true" maxheight="auto" offsety="true" origin="top left" transition="v-menu-transition">
  <vlist-stub tag="div"></vlist-stub>
</vmenu-stub>
`;
